src/read/elf/version.rs

```diff
@@ -293,6 +293,134 @@ impl<'data, Elf: FileHeader> VerdauxIterator<'data, Elf> {
     }
 }
 
+/// An iterator over the section entries in an ELF `SHT_GNU_attributes` section.
+#[derive(Debug, Clone)]
+pub struct AttribSectionIterator<'data, Elf: FileHeader> {
+    endian: Elf::Endian,
+    data: Bytes<'data>,
+}
+
+impl<'data, Elf: FileHeader> AttribSectionIterator<'data, Elf> {
+    pub(super) fn new(endian: Elf::Endian, data: Bytes<'data>) -> Self {
+        AttribSectionIterator { endian, data: data }
+    }
+
+    /// Return the next Vendor attribute section
+    pub fn next(
+        &mut self,
+    ) -> Result<
+        Option<(
+            endian::U32<Elf::Endian>,
+            &'data [u8],
+            TagIterator<'data, Elf>,
+        )>,
+    > {
+        if self.data.is_empty() {
+            return Ok(None);
+        }
+
+        // Don't move self.data so that we can simply add section_length
+        // to it to move to the next vendor section
+        let mut data = self.data;
+        let section_length = *data
+            .read::<endian::U32<Elf::Endian>>()
+            .read_error("ELF GNU attributes vendor section is too short")?;
+        let vendor_name = data
+            .read_string()
+            .read_error("ELF GNU attributes vendor section is too short")?;
+
+        let tags = TagIterator::new(
+            self.endian,
+            data.read_bytes(
+                // calculate the remaining space in this section
+                // section - size field - vendor name string - NULL terminator
+                section_length.get(self.endian) as usize
+                    - core::mem::size_of::<endian::U32<Elf::Endian>>()
+                    - vendor_name.len()
+                    - 1,
+            )
+            .read_error("ELF GNU attributes vendor section is too short")?,
+        );
+
+        let next = section_length.get(self.endian);
+        if next != 0 {
```

Comment:
I tried to do this everywhere, there is only one instance of it remaining. In `AttribSubSubsectionIterator::next()` I still have some arithmetic  where I have to reduce the remaining size by the tag and length fields we already read. I kept it this way since otherwise I would read the tag/size, read the entire subsection based on that size, skip the tag/size a second time, and then use the remaining for the attribute data. The current way avoids having to do everything twice like that, but let me know if you have a better idea in mind you'd like me to do.